DOC: Clarify the type alias deprecation message (#18230)
numpy/numpy: doc/source/release/1.20.0-notes.rst

----------------------------------------------------------------

For a long time, ``np.int`` has been an alias of the builtin ``int``. This is
repeatedly a cause of confusion for newcomers, and existed mainly for historic
reasons.

These aliases have been deprecated. The table below shows the full list of
deprecated aliases, along with their exact meaning. Replacing uses of items in
the first column with the contents of the second column will work identically
and silence the deprecation warning.

The third column lists alternative NumPy names which may occasionally be

> preferential. See also `user/basics.types`_ for additional details.

Reviewer: The reference is giving an error.

=================  ============  ==================================================================
Deprecated name    Identical to  NumPy scalar type names
=================  ============  ==================================================================
``numpy.bool``     ``bool``      `numpy.bool_`
``numpy.int``      ``int``       `numpy.int_` (default), ``numpy.int64``, or ``numpy.int32``
``numpy.float``    ``float``     `numpy.float64`, `numpy.float_`, `numpy.double` (equivalent)

> ``numpy.complex``  ``complex``   `numpy.complex128`, ``numpy.complex_`, `numpy.cdouble` (equivalent)

Reviewer: Needs backtick  `numpy.complex_`. Or maybe fewer :)

``numpy.object``   ``object``    `numpy.object_`
``numpy.str``      ``str``       `numpy.str_`
``numpy.long``     ``int``       `numpy.int_` (C ``long``), `numpy.longlong` (largest integer type)
``numpy.unicode``  ``str``       `numpy.unicode_`
=================  ============  ==================================================================

To give a clear guideline for the vast majority of cases, for the types
``bool``, ``object``, ``str`` (and ``unicode``) using the plain version
is shorter and clear, and generally a good replacement.
For ``float`` and ``complex`` you can use ``float64`` and ``complex128``
if you wish to be more explicit about the precision.

> For ``np.int`` a direct replacement with ``np.int_`` or ``int `` is also

Reviewer: Something is causing a problem here. Maybe the space after `int`?

> good, but the precision depends on the computer and operating system.

Reviewer 2: Perhaps this isn't the place to say it, but I think it's important to note that `np.int_` and `int` are interpreted in the same way, even though that way is platform-dependent. This wording suggests that these replacements are platform dependent while the original was not, which is misleading.
```suggestion
good and will not change behavior, but the precision will continue to depend
on the computer and operating system.
```

If you want to be more explicit and review the current use, you have the
following alternatives:

* ``np.int64`` or ``np.int32`` to specify the precision exactly.
  This ensures that results cannot depend on the computer or operating system.
* ``np.int_`` or ``int`` (the default), but be aware that it depends on
  the computer and operating system.
* The C types: ``np.cint`` (int), ``np.int_`` (long), ``np.longlong``,
* ``np.intp`` which is 32bit on 32bit machines 64bit on 64bit machines.
  This can be the best type to use for indexing.

When used with ``np.dtype(...)`` or ``dtype=...`` changing it to the
NumPy name as mentioned above will have no effect on the output.
If used as a scalar with::

    np.float(123)

changing it can subtly change the result.  In this case, the Python version
``float(123)`` or ``int(12.)`` is normally preferable, although the NumPy

> version may be useful for consistency with NumPy arrays. For example, since
> NumPy behaves differently for things like division by zero.

Reviewer 2: Last sentence feels incomplete to me, perhaps better as:
```suggestion
version may be useful for consistency with NumPy arrays (for example,
NumPy behaves differently for things like division by zero).
```

``numpy.str``      ``str``       `numpy.str_`
``numpy.long``     ``int``       `numpy.int_` (C ``long``), `numpy.longlong` (largest integer type)
``numpy.unicode``  ``str``       `numpy.unicode_`
=================  ============  ==================================================================

To give a clear guideline for the vast majority of cases, for the types
``bool``, ``object``, ``str`` (and ``unicode``) using the plain version
is shorter and clear, and generally a good replacement.
For ``float`` and ``complex`` you can use ``float64`` and ``complex128``
if you wish to be more explicit about the precision.

For ``np.int`` a direct replacement with ``np.int_`` or ``int`` is also
good and will not change behavior, but the precision will continue to depend
on the computer and operating system.
If you want to be more explicit and review the current use, you have the
following alternatives:

* ``np.int64`` or ``np.int32`` to specify the precision exactly.
  This ensures that results cannot depend on the computer or operating system.
* ``np.int_`` or ``int`` (the default), but be aware that it depends on
  the computer and operating system.

> * The C types: ``np.cint`` (int), ``np.int_`` (long), ``np.longlong``,

Author: ```suggestion
* The C types: ``np.cint`` (int), ``np.int_`` (long), ``np.longlong``.
```
